include/list.h

```diff
@@ -0,0 +1,406 @@
+ *
+ * The function used to compare the elements of the liste when doing
+ * operations on an ordered list.
+ * @param data1 - First element to be compared
+ * @param data2 - Second element to be compared
+ * @return
+ *  - -1 - If data1 < data2
+ *  - 0 - If data1 == data2
+ *  - 1 - If data1 > data2
+ */
+typedef int32_t (*f_cmp)(void *data1, void *data2);
+
+/**
+ * @name Generic functions
+ * Each function interacting with the list have one of the following formats.\n
+ * Aditionaly they may have one more parametere for specific functionalities.\n
+ * In the Iterator functions, the list reference is replaced by the iterator's
+ * one.
+ * @{
+ */
+
+/**
+ * @brief Add an element in the list.
+ *
+ * The element of the list is created and the data field is stored in it.
+ * @param list_desc - Reference to the list. Created by \ref list_init
+ * @param data - Data to store in a list element
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+typedef int32_t	(*f_add)(List_desc list_desc, void *data);
+
+/**
+ * @brief Edit an element in the list. The content is replaced by new_data.
+ * @param list_desc - Reference to the list. Created by \ref list_init .
+ * @param new_data - New data to replace the old one
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+typedef int32_t	(*f_edit)(List_desc list_desc, void *new_data);
+
+/**
+ * @brief Read an element from the list.
+ * @param list_desc - Reference to the list. Created by \ref list_init
+ * @param result - If not null, result is filled with:
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ * @return \n
+ *  - Content of the list element
+ *  - NULL : If some error occure
+ * @note If the content of an element can be 0 then the result must be checked
+ * to see if the functions has succeded
+ */
+typedef void*	(*f_read)(List_desc list_desc, int32_t *result);
+
+/**
+ * @brief Read and remove an element from the list.
+ * @param list_desc - Reference to the list. Created by \ref list_init
+ * @param result - If not null, result is filled with:
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ * @return
+ *  - Content of the list element
+ *  - NULL : If some error occure
+ * @note If the content of an element can be 0 then the result must be checked
+ * to see if the functions has succeded
+ */
+typedef void*	(*f_get)(List_desc list_desc, int32_t *result);
+
+/** @} */
+
+/**
+ * @enum adapter_type
+ * @brief Posibles adapters for the implemented list
+ */
+enum adapter_type {
+	/** Default type. The created adapter will be like for LIST_STACK */
+	LIST_DEFAULT,
+	/**
+	 * Adapter for a FIFO list (Last-in first-out). Elements are inserted
+	 * and extracted only from one end
+	 *  - \e Push: Insert element
+	 *  - \e Pop: Get next element (Read and remove)
+	 *  - \e Top_next: Read next element
+	 *  - \e Back: Read first element
+	 *  - \e Swap: Edit the content of the next element
+	 */
```

Comment:
Comment is a bit misleading. FIFO = first in first out. This would be the correct implementation of a queue as the name suggests, but elements are inserted through one end and extracted from the other.

Reply:
solved